OsmAnd/src/net/osmand/plus/plugins/aistracker/AisObject.java

```diff
@@ -988,4 +1006,121 @@ public float getDistanceInNauticalMiles() {
     public boolean getSignalLostState() {
         return (isLost(vesselLostTimeoutInMinutes) && isMovable() && !vesselAtRest);
     }
+
+    private float getVesselRotation()
+    {
+        float rotation = 0;
+        if (this.ais_cog != INVALID_COG) { rotation = (float)this.ais_cog; }
+        else if (this.ais_heading != INVALID_HEADING ) { rotation = this.ais_heading; }
+        return rotation;
+    }
+
+    public void createAisRenderData(@NonNull Context context, int baseOrder,
+                                    @NonNull AisTrackerLayer mapLayer, @NonNull Paint paint,
+                                    @NonNull MapMarkersCollection markersCollection,
+                                    @NonNull VectorLinesCollection vectorLinesCollection) {
+        updateBitmap(mapLayer, paint);
+
+        assert bitmap != null;
+
+        float CircleCIDensity = 5;
+        ChartPointsHelper chartPointsHelper = new ChartPointsHelper(context);
```

Comment:
Do we need it only for V2? better store in layer instead of recreating for every AisObject

Reply:
Moved to layer